# Problemas de prueba de hipótesis

## Ejercicio 1

Usted es un nutricionista que investiga dos tipos diferentes de dietas para ver si existe una diferencia significativa en la pérdida de peso después de un mes. Eliges dos grupos aleatorios de personas; un grupo sigue la primera dieta y el otro sigue la segunda. Al final del mes, se registra la pérdida de peso (en kg) de cada persona:

| Diet 1 | Diet 2 |
|:-------|:-------|
| 2.0 | 3.0 |
| 2.5 | 3.2 |
| 3.0 | 3.1 |
| 2.8 | 2.9 |
| 2.3 | 2.8 |
| 2.7 | 3.0 |
| 2.5 | 3.2 |

Con estos datos se busca responder a la siguiente pregunta: ¿Existe una diferencia significativa en la pérdida de peso promedio entre las personas que siguieron la primera dieta y las que siguieron la segunda dieta?

Para sacar conclusiones, siga los siguientes puntos:

-Enunciar la hipótesis: hipótesis nula y alternativa.
-Realizar la prueba para comprobar la hipótesis. Puede utilizar la prueba t de Student.
-Analizar las conclusiones.

In [ ]:
import numpy as np
from scipy import stats

# Definimos los datos
dieta1 = [2.0, 2.5, 3.0, 2.8, 2.3, 2.7, 2.5]
dieta2 = [3.0, 3.2, 3.1, 2.9, 2.8, 3.0, 3.2]

# Primero debemos comprobar que los grupos deben seguir una distribución normal
print('P_Valor1 = ',stats.shapiro(dieta1).pvalue)
print('P_Valor2 = ',stats.shapiro(dieta2).pvalue)

# Ahora debemos comprobar si la varianza entre grupos es la misma
print('P_Valor = ',stats.levene(dieta1, dieta2).pvalue)

P_Valor1 =  0.9635409177367069
P_Valor2 =  0.5905246580529597
P_Valor =  0.15767770055358


Puesto que los P_Valores > 0.05, se concluye que los datos son consistentes con una Distribución Normal y no se detecta una diferencia significativa entre las varianzas de los dos grupos.

Continuamos con el análisis

In [ ]:
# TODO
# Enunciamos las hippotesís
H0 = 'No hay diferencia significativa en la pérdida de peso promedio de las dos dietas.'
H1 = 'Hay diferencia significativa en la pérdida de peso promedio de las dos dietas.'

print('Hipotesis Nula =', H0,'\nHipotesis Alternativa =', H1,'\n')

# Realizamos la prueba t de Student
t_statistico, p_value = stats.ttest_ind(dieta1, dieta2)

# Imprimimos los resultados
print(f"Estadístico t: {t_statistico}")
print(f"Valor p: {p_value}\n")

alpha = 0.05

if p_value < alpha:
    print("La evidencia de los datos es suficientemente fuerte como para rechazar la hipótesis nula H₀, es decir, respaldan la Hipótesis alternativa (H₁).")
    print(f"Es muy improbable que la evidencia observada se deba al azar (p-valor={p_value}).")
else:
    print("No hay suficiente evidencia para rechazar la hipótesis nula H₀, es decir, no hay suficiente evidencia para afirmar que la hipótesis nula es falsa.")
    print("Esto no implica que aceptemos H₀, ya que la ausencia de evidencia no es lo mismo que evidencia en su favor.")
    print(f"Es muy probable que la evidencia observada se deba al azar (p-valor={p_value}).")

# Calculamos las medias de cada dieta
media_dieta1 = np.mean(dieta1)
media_dieta2 = np.mean(dieta2)

print(f"\nMedia de pérdida de peso en Dieta 1: {media_dieta1:.2f} kg")
print(f"\nMedia de pérdida de peso en Dieta 2: {media_dieta2:.2f} kg")

Hipotesis Nula = No hay diferencia significativa en la pérdida de peso promedio de las dos dietas. 
Hipotesis Alternativa = Hay diferencia significativa en la pérdida de peso promedio de las dos dietas. 

Estadístico t: -3.5383407969933938
Valor p: 0.004083270191713912

La evidencia de los datos es suficientemente fuerte como para rechazar la hipótesis nula H₀, es decir, respaldan la Hipótesis alternativa (H₁).
Es muy improbable que la evidencia observada se deba al azar (p-valor=0.004083270191713912).

Media de pérdida de peso en Dieta 1: 2.54 kg

Media de pérdida de peso en Dieta 2: 3.03 kg


## ANOVA

**ANOVA** (*Analysis of Variance - Análisis de variación*) Es una técnica estadística utilizada para comparar las medidas de dos o más grupos. La idea detrás de ANOVA es descomponer la variabilidad total de los datos en dos componentes: variabilidad entre grupos y variabilidad dentro de los grupos:

- **Variabilidad entre grupos**: Esta variabilidad se refiere a las diferencias entre las medias del grupo. Si esta variabilidad es considerablemente mayor que la variabilidad dentro del grupo, podría ser una indicación de que al menos una de las medias del grupo es diferente.
- **Variabilidad dentro del grupo**: Esta variabilidad se refiere a la dispersión de los datos dentro de cada grupo. Si todos los grupos tienen una variabilidad similar, entonces cualquier diferencia notable en las medias de los grupos podría considerarse significativa.

Las hipótesis en ANOVA suelen incluir:

- **Hipótesis nula** ($H₀$): Las medias de todos los grupos son iguales.
- **Hipótesis alternativa** ($H₁$): Al menos una de las medias del grupo es diferente.

Si el resultado de la prueba ANOVA es significativo (por ejemplo, un valor p inferior a un umbral como 0,05), esto sugiere que al menos la media de un grupo es diferente.

## Ejercicio 2

Un agricultor decide probar tres tipos diferentes de fertilizantes para determinar si alguno es superior en términos de producción de maíz. El agricultor planta maíz en 15 parcelas idénticas y utiliza los tres fertilizantes (5 parcelas para cada tipo). Al final de la temporada mide el rendimiento de maíz (en kg) de cada parcela, con el siguiente resultado:

| Fertilizante 1 | Fertilizante 2 | Fertilizante 3 |
|:-------------|:-------------|:-------------|
| 20 | 22 | 24 |
| 21 | 21 | 23 |
| 20 | 23 | 22 |
| 19 | 22 | 23 |
| 20 | 21 | 24 |

Con estos datos busca responder la siguiente pregunta: ¿Existe una diferencia significativa en el rendimiento promedio del maíz entre los tres tipos de fertilizantes?

Para ayudarte, sigue los siguientes puntos:

-Enunciar la hipótesis: hipótesis nula y alternativa.
-Realizar la prueba ANOVA.
-Analizar las conclusiones.
-Si un fertilizante es mejor que otro, ¿cómo podemos saberlo?

In [32]:
# Definimos los datos
fertilizante1 = [20, 21, 20, 19, 20]
fertilizante2 = [22, 21, 23, 22, 21]
fertilizante3 = [24, 23, 22, 23, 24]

# Primero debemos comprobar que los grupos deben seguir una distribución normal
print('P_Valor1 = ',stats.shapiro(fertilizante1).pvalue)
print('P_Valor2 = ',stats.shapiro(fertilizante2).pvalue)
print('P_Valor3 = ',stats.shapiro(fertilizante3).pvalue)

# Ahora debemos comprobar si la varianza entre grupos es la misma
print('P_Valor = ',stats.levene(fertilizante1, fertilizante2, fertilizante3).pvalue)

P_Valor1 =  0.32542950716414576
P_Valor2 =  0.3140395561440371
P_Valor3 =  0.3140395561440371
P_Valor =  0.8039599174006208


Puesto que los P_Valores > 0.05, se concluye que los datos son consistentes con una Distribución Normal y no se detecta una diferencia significativa entre las varianzas de los tres grupos.

Continuamos con el análisis

In [33]:
# TODO
# Enunciamos las hipótesis
print('Hipótesis nula (H0): No hay diferencia significativa en el rendimiento promedio del maíz entre los tres tipos de fertilizantes.\n')
print('Hipótesis alternativa (H1): Al menos un fertilizante tiene un rendimiento promedio de maíz significativamente diferente.\n')

# Realizamos la prueba ANOVA
f_statistico, p_valor = stats.f_oneway(fertilizante1, fertilizante2, fertilizante3)

# Imprimimos los resultados
print(f"Estadístico F: {f_statistico}")
print(f"Valor p: {p_valor}\n")

alpha = 0.05

if p_valor < alpha:
    print("La evidencia de los datos es suficientemente fuerte como para rechazar la hipótesis nula H₀, es decir, respaldan la Hipótesis alternativa (H₁).")
    print(f"Es muy improbable que la evidencia observada se deba al azar (p-valor={p_valor}).")
else:
    print("No hay suficiente evidencia para rechazar la hipótesis nula H₀, es decir, no hay suficiente evidencia para afirmar que la hipótesis nula es falsa.")
    print("Esto no implica que aceptemos H₀, ya que la ausencia de evidencia no es lo mismo que evidencia en su favor.")
    print(f"Es muy probable que la evidencia observada se deba al azar (p-valor={p_valor}).")

media_fert_1 = np.mean(fertilizante1)
media_fert_2 = np.mean(fertilizante2)
media_fert_3 = np.mean(fertilizante3)

print(f"\nMedia de peso en Fertilizante1: {media_fert_1:.2f} kg")
print(f"Media de peso en Fertilizante2: {media_fert_2:.2f} kg")
print(f"Media de peso en Fertilizante3: {media_fert_3:.2f} kg")

Hipótesis nula (H0): No hay diferencia significativa en el rendimiento promedio del maíz entre los tres tipos de fertilizantes.

Hipótesis alternativa (H1): Al menos un fertilizante tiene un rendimiento promedio de maíz significativamente diferente.

Estadístico F: 20.315789473684188
Valor p: 0.00014047824793190475

La evidencia de los datos es suficientemente fuerte como para rechazar la hipótesis nula H₀, es decir, respaldan la Hipótesis alternativa (H₁).
Es muy improbable que la evidencia observada se deba al azar (p-valor=0.00014047824793190475).

Media de peso en Fertilizante1: 20.00 kg
Media de peso en Fertilizante2: 21.80 kg
Media de peso en Fertilizante3: 23.20 kg


El test de Tukey nos indicará específicamente qué pares de grupos tienen diferencias significativas.

In [ ]:
import statsmodels.api as sm
from statsmodels.stats.multicomp import pairwise_tukeyhsd

# Concatenamos los datos y creamos etiquetas
rendimientos = np.concatenate([fertilizante1, fertilizante2, fertilizante3])
etiquetas = ['Fertilizante 1'] * len(fertilizante1) + ['Fertilizante 2'] * len(fertilizante2) + ['Fertilizante 3'] * len(fertilizante3)

# Realizamos el test de Tukey
tukey_resultados = pairwise_tukeyhsd(rendimientos, etiquetas)

# Imprimimos los resultados del test de Tukey
print('\ntukey_resultados:\n',tukey_resultados)


tukey_resultados:
        Multiple Comparison of Means - Tukey HSD, FWER=0.05        
    group1         group2     meandiff p-adj  lower  upper  reject
------------------------------------------------------------------
Fertilizante 1 Fertilizante 2      1.8 0.0099 0.4572 3.1428   True
Fertilizante 1 Fertilizante 3      3.2 0.0001 1.8572 4.5428   True
Fertilizante 2 Fertilizante 3      1.4 0.0409 0.0572 2.7428   True
------------------------------------------------------------------


Conclusiones del test de tukey:

a) Fertilizante 1 vs Fertilizante 2:
Diferencia media: 1.8
Valor p ajustado: 0.0099 (< 0.05)
Se rechaza la hipótesis nula
Conclusión: Hay una diferencia significativa entre estos fertilizantes.

b) Fertilizante 1 vs Fertilizante 3:
Diferencia media: 3.2
Valor p ajustado: 0.0001 (< 0.05)
Se rechaza la hipótesis nula
Conclusión: Hay una diferencia muy significativa entre estos fertilizantes.

c) Fertilizante 2 vs Fertilizante 3:
Diferencia media: 1.4
Valor p ajustado: 0.0409 (< 0.05)
Se rechaza la hipótesis nula
Conclusión: Hay una diferencia significativa entre estos fertilizantes.

Conclusión General: Existen diferencias significativas entre todos los pares de fertilizantes. El Fertilizante 3 parece ser el más efectivo, seguido por el Fertilizante 2, y el Fertilizante 1 es el menos efectivo. La mayor diferencia se observa entre el Fertilizante 1 y el Fertilizante 3.